Emit a warning when let override a method defined in the same context

fabn opened the pull request:
Refs #1901

myronmarston commented on lib/rspec/core/memoized_helpers.rb:
```diff
@@ -232,6 +232,9 @@ def let(name, &block)
           # allow it to use method constructs like `super` and `return`.
           raise "#let or #subject called without a block" if block.nil?
+          if method_defined?(name) && instance_method(name).owner == self
+            RSpec.warning "method :#{name} defined multiple times through let in the same context"
```
This warning string is accurate for your case and the specs you wrote below, but there are cases where it would be emitted and would not be accurate:

``` ruby
RSpec.describe do
  def foo; end
  let(:foo) { }
end
```

In such a situation, `let(:foo)` is stomping `def foo` but the warning describes both as `let`.  I also think that this warning could be a bit clearer in general -- it says there are multiple `let(#{name})` but doesn't explain why that is problematic.  How about this instead?

``` ruby
RSpec.warning "`#{self.name}##{name}` is being redefined through the `let(:#{name})` " \
              "definition at #{RSpec::CallerFilter.first_non_rspec_line}. The original " \
              "definition will never be used and can be removed", :call_site => nil
```

...which would look like:

```
WARNING: `RSpec::ExampleGroups::Anonymous#foo` is being redefined through the `let(:foo)` definition at /Users/myron/code/rspec-dev/repos/rspec-core/spec/rspec/core/memoized_helpers_spec.rb:436:in `block (3 levels) in <module:Core>'. The original definition will never be used and can be removed.
```

I think that's more clear.  WDYT?

myronmarston replied:
BTW, it would be nice to add a spec for the `def foo; end; let(:foo)` case as well.  One can imagine an implementation that tracked `let` defs directly and warned on a `let` override of a prior `let` did not warn when it overrode a `def`.

fabn replied:
Your warning message is definitely better, I'll update it.

myronmarston commented on spec/rspec/core/memoized_helpers_spec.rb:
```diff
@@ -421,6 +421,23 @@ def count
+          let(:foo) { 'second value' }
+        end
+      end
+    end
```
Instead of `silence_warnings { expect(RSpec).to receive(:warning); ... }`, WDYT of this?

``` ruby
it 'emits a warning when declared multiple times' do
  line = nil
  expect {
    RSpec.describe do
      let(:foo) { 'first value' }
      let(:foo) { 'second value' }; line = __LINE__
    end
  }.to output(a_string_including("#foo", "let(:foo)", "definition at #{__FILE__}:#{line}")).to_stderr
end
```

I think it's better because it doesn't require `silence_warnings` and it's less coupled to the implementation -- it simply passes if the snippet is written to stderr, regardless of how it got there (e.g. via `Kernel#warn` or `RSpec.warning` or whatever).  `a_string_including` lets you easily match multiple snippets, including "definition at #{line}" which is important so that the user is notified of the line where the problem is.

myronmarston commented on spec/rspec/core/memoized_helpers_spec.rb:
```diff
@@ -429,6 +446,16 @@ def count
+          context 'in sub context' do
+            let(:foo) { 'second value' }
+          end
+        end
```
As above, I think it's better to use the `output` matcher here:

``` ruby
it 'does not emit warnings on override' do
  expect {
    RSpec.describe do
      let(:foo) { 'first value' }
      context 'in sub context' do
        let(:foo) { 'second value' }
      end
    end
  }.to avoid_outputting.to_stdout.and avoid_outputting.to_stderr
end
```

* `avoid_outputting` is a negated form of `output`: https://github.com/rspec/rspec-core/blob/1e483f6de33a56bb93d8846644cf73e42d307596/spec/support/matchers.rb#L112
* If the implementation switched to using `Kernel#warn` (or some other mechanism), then the use of `expect(RSpec).to_not receive(:warning)` would be a false positive.  Really, we care that no output is emitted to stdout/stderr when loading this example group, regardless of the mechanisms for doing so.

myronmarston commented:
This is a looking good, @fabn, but I do have a couple thoughts of potential issues:

* This will notice when there are duplicate `let` defs, but not when there are duplicate method defs (via `def` or `define_method`), which could also be a problem with using `include_context` twice in the same group as in your original issue.
* For the `include_context` case, the user may not be aware that using it twice in the same context is prone to this issue and the simple work-around of using `it_behaves_like` instead.  It would be nice to provide them a better warning in this case to direct the user to the solution.

For the first issue, I think we can solve it by using a `method_added` hook instead of the logic you've put in place for `let`:

http://ruby-doc.org/core-2.2.0/Module.html#method-i-method_added

For the latter issue, we'd have to make the logic aware of whether or not it is currently evaluating a shared example group block via `include_examples`. We do some tracking of that (in order to provide the shared group inclusion backtrace on failures) but I don't think that is necessarily directly usable by this.   We'd probably have to add a bit more code to support that.  I'm not sure if it's worth the effort -- we can only do so much to provide good warnings.  In a turing-complete language the user is always going to be able to do things that may confuse them and we can't prevent it.

WDYT?

myronmarston commented on features/example_groups/shared_examples.feature:
````diff
@@ -19,6 +19,46 @@ Feature: shared examples
+  ```
+
+  To prevent this kind of subtle errors a warning is raised if you declare multiple
+  let blocks with the same name in the same context.
````
Well done on the doc update...it definitely helps.  Do you think we should mention the solution being to use `it_behaves_like`, which uses nesting to avoid the issue?

Also, two minor nitpicks:

*  "warning is raised" is perhaps a bit confusing -- `raise` has the specific meaning in ruby of raising an exception.  Maybe "a warning is emitted" would avoid that confusion?
* Would be nice to put backticks around `let` so it renders as code on relish.

myronmarston replied:
Also, s/subtle errors/subtle error/ (reads better as singular, I think).

fabn commented:
> This will notice when there are duplicate let defs, but not when there are duplicate method defs (via def or define_method), which could also be a problem with using include_context twice in the same group as in your original issue.

I think that majority of users don't use `def` or `define_method` inside `shared_(example|behavior)`, they don't belong to RSpec DSL and should be in some way avoided. However since it's possible to do that a `method_added` hook will catch all of these way to add methods and it may be a better way to do the check.

> For the first issue, I think we can solve it by using a method_added hook instead of the logic you've put in place for let:

I'll try to experiment with `method_added`, it's possible that we get the intended behavior for free just moving the check there.

> For the include_context case, the user may not be aware that using it twice in the same context is prone to this issue and the simple work-around of using it_behaves_like instead. It would be nice to provide them a better warning in this case to direct the user to the solution.

> For the latter issue, we'd have to make the logic aware of whether or not it is currently evaluating a shared example group block via include_examples. We do some tracking of that (in order to provide the shared group inclusion backtrace on failures) but I don't think that is necessarily directly usable by this. We'd probably have to add a bit more code to support that. I'm not sure if it's worth the effort -- we can only do so much to provide good warnings.

I agree with you, adding code for this is too much effort for just a warning.

> In a turing-complete language the user is always going to be able to do things that may confuse them and we can't prevent it.

:+1: :+1:  :+1:

fabn commented:
> For the first issue, I think we can solve it by using a method_added hook instead of the logic you've put in place for let:
> I'll try to experiment with method_added, it's possible that we get the intended behavior for free just moving the check there.

It's impossible to use `method_added` to do the check, because when it's executed the added method is already part of the class so we have 

```ruby 
method_defined?(name) # => always true
instance_method(name).owner == self # => always true
```

fabn commented:
Ugly way to detect override on self with `method_added` (it actually works)

```ruby
def method_added(name)
  @count ||= {}
  @count[name] ||= 0
  @count[name] += 1
  if ancestors.select { |k| k.method_defined?(name) } == [self] && @count[name] > 1
    RSpec.warning ...
  end
end
```

I don't think you want me to commit this kind of code, don't you?

myronmarston commented:
There's a far simpler way to do it with `method_added`. Add this class method to `RSpec::Core::ExampleGroup`:

``` ruby
 def self.method_added(method_name)
  if (@added_methods ||= Set.new).include?(method_name)
    RSpec.warning "`#{name}##{method_name}` is being redefined " \
                  "at #{RSpec::CallerFilter.first_non_rspec_line}. The original " \
                  "definition will never be used and can be removed", :call_site => nil
  end

  @added_methods << method_name
end
```

myronmarston commented on spec/rspec/core/memoized_helpers_spec.rb:
```diff
@@ -421,6 +421,26 @@ def count
+          let(:foo) { 'first value' }
+          let(:foo) { 'second value' }; line = __LINE__
+        end
+      }.to output(a_string_including("#foo", "let(:foo)", "definition at #{__FILE__}:#{line}")).to_stderr
```
So I realized something about our use of `line` here...when the `a_string_including` matcher is created, `line` is `nil` because the `expect` block hasn't executed yet.  So the last arg to `a_string_including` is the same as `"definition at #{__FILE__}:#{nil}"` which passes because `nil.to_s == ""`.  I think we're better off using `#{__LINE__ - 2}` instead of a `line` local to avoid this issue.

fabn commented:
> There's a far simpler way to do it with method_added. Add this class method to RSpec::Core::ExampleGroup:

That would work even if not thread safe (as everything that deal with class variables with no synchronization). Should I also move specs in `spec/rspec/core/example_group_spec.rb`?

fabn commented:
Pushed some more commit. Now it should work. Let me know if it's fine for you and if I should squash the PR for merging.

fabn commented:
Mmm, got a [rubocop error](https://travis-ci.org/rspec/rspec-core/jobs/54570873) about class length.

I don't know how to fix (changing rubocop config, use a module, what else?), please advise.

myronmarston commented on lib/rspec/core/example_group.rb:
```diff
@@ -650,6 +650,16 @@ def self.method_missing(name, *args)
 
+      # @private
+      def self.method_added(method_name)
+        if (@added_methods ||= Set.new).include?(method_name)
```
Given that users can write class methods (e.g. macros) in their example groups, and such methods would run with access to the `@added_methods`, I'd prefer to see this use `@__added_methods` to reduce the likelihood of naming conventions.

myronmarston commented:
> Mmm, got a rubocop error about class length.
> 
> I don't know how to fix (changing rubocop config, use a module, what else?), please advise.

Let's just exclude `ExampleGroup` from that check.  It's naturally a long class because it supports so much of RSpec's functionality.  I don't think that check is valuable on that class.  Here's how to do that:

https://github.com/rspec/rspec-core/blob/12250327fe17c87a142e8fe5b607a52e3d71e2bc/lib/rspec/core/configuration.rb#L7
https://github.com/rspec/rspec-core/blob/12250327fe17c87a142e8fe5b607a52e3d71e2bc/lib/rspec/core/configuration.rb#L1760

Notice the blank line between the comment and the doc comments -- that way we keep the rubocop comment from showing up on the docs.

> That would work even if not thread safe (as everything that deal with class variables with no synchronization). Should I also move specs in spec/rspec/core/example_group_spec.rb?

:+1:

I'm not worried about threadsafety here.  We control the loading of spec files, not users, and do so in single-threaded manner.

> Let me know if it's fine for you and if I should squash the PR for merging.

I left one suggestion about changing a variable name, but otherwise this is good to guy and I'd appreciate it being squashed.  Thanks for the great work!

fabn commented:
> I left one suggestion about changing a variable name, but otherwise this is good to guy and I'd appreciate it being squashed. 

Rebased and squashed.

> Thanks for the great work!

Thank you, it was a pleasure to work with a such great project.
